BitwardenShared/UI/Auth/Utilities/VaultUnlockSetupHelper.swift

```diff
@@ -0,0 +1,149 @@
+            showAlert(.enterPINCode(
+                onCancelled: {
+                    continuation.resume(throwing: VaultUnlockSetupHelperError.userCancelled)
+                },
+                completion: { pin in
+                    continuation.resume(returning: pin)
+                }
+            ))
+        }
+    }
+
+    /// Shows an alert asking the user whether they want to require their master password to unlock
+    /// the vault when the app restarts.
+    ///
+    /// - Parameters:
+    ///   - biometricType: The biometric type the app supports.
+    ///   - showAlert: A closure used to handle showing the alert.
+    /// - Returns: Whether the user wants to require their master password on app restart.
+    ///
+    private func showUnlockWithPinAlert(
+        biometricType: BiometricAuthenticationType?,
+        showAlert: @escaping (Alert) -> Void
+    ) async -> Bool {
+        await withCheckedContinuation { continuation in
+            showAlert(.unlockWithPINCodeAlert(biometricType: biometricType) { requirePassword in
+                continuation.resume(returning: requirePassword)
+            })
+        }
+    }
+}
+
+// MARK: - DefaultVaultUnlockSetupHelper+VaultUnlockSetupHelper
+
+extension DefaultVaultUnlockSetupHelper: VaultUnlockSetupHelper {
+    func setBiometricUnlock(enabled: Bool, showAlert: @escaping (Alert) -> Void) async -> BiometricsUnlockStatus? {
+        do {
+            try await services.authRepository.allowBioMetricUnlock(enabled)
+            return try await services.biometricsRepository.getBiometricUnlockStatus()
+        } catch {
+            services.errorReporter.log(error: error)
+            showAlert(.defaultAlert(title: Localizations.anErrorHasOccurred))
+            return try? await services.biometricsRepository.getBiometricUnlockStatus()
+        }
+    }
+
+    func setPinUnlock(enabled: Bool, showAlert: @escaping (Alert) -> Void) async -> Bool {
+        do {
+            if enabled {
+                let pin = try await showEnterPinAlert(showAlert: showAlert)
+                guard !pin.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
+
+                let userHasMasterPassword = try await services.stateService.getUserHasMasterPassword()
+                let biometricType = services.biometricsRepository.getBiometricAuthenticationType()
+                let requirePasswordAfterRestart = if userHasMasterPassword {
+                    await showUnlockWithPinAlert(biometricType: biometricType, showAlert: showAlert)
+                } else {
+                    false
+                }
+
+                try await services.authRepository.setPins(
+                    pin,
+                    requirePasswordAfterRestart: requirePasswordAfterRestart
+                )
+            } else {
+                try await services.authRepository.clearPins()
+            }
+
+            return enabled
```

Comment:
⛏️ Could we use guard to reduce nesting?
```suggestion
            guard enabled else {
                try await services.authRepository.clearPins()
                return false
            }

            let pin = try await showEnterPinAlert(showAlert: showAlert)
            guard !pin.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }

            let userHasMasterPassword = try await services.stateService.getUserHasMasterPassword()
            let biometricType = services.biometricsRepository.getBiometricAuthenticationType()
            let requirePasswordAfterRestart = if userHasMasterPassword {
                await showUnlockWithPinAlert(biometricType: biometricType, showAlert: showAlert)
            } else {
                false
            }

            try await services.authRepository.setPins(
                pin,
                requirePasswordAfterRestart: requirePasswordAfterRestart
            )
            return true
```